### DISCUSSION TITLE
Error with npm on ubuntu: executable file not found in $PATH

### DISCUSSION THREAD

**Discussion by adamtrip** [microsoft/aspire#5859] [General]
Hi! Having this issue on aspire with npm app.

```
fail: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      failed to start a process {"Executable": {"name":"frontend-mrxuuacr"}, "Reconciliation": 8, "error": "exec: \"npm\": executable file not found in $PATH"}
fail: Aspire.Hosting.Dcp.dcpctrl.ExecutableReconciler[0]
      failed to start Executable        {"Executable": {"name":"frontend-mrxuuacr"}, "Reconciliation": 8, "error": "exec: \"npm\": executable file not found in $PATH"}

```

this is my config:
```
builder.AddNpmApp("frontend", "./../../dm-studio-frontend", "dev")
    .WithHttpEndpoint(port: 3000, env: "PORT")
    .WithReference(backend)
    .PublishAsDockerFile();
```

I also tried to create a starter app, creating a new react app and adding a similar config as above:
```
var builder = DistributedApplication.CreateBuilder(args);

var apiService = builder.AddProject<Projects.testing_ApiService>("apiservice");

builder.AddProject<Projects.testing_Web>("webfrontend")
    .WithExternalHttpEndpoints()
    .WithReference(apiService);

builder.AddNpmApp("fr", "../my-app", "dev");

builder.Build().Run();
```

Screenshot of the dashboard:

![image](https://github.com/user-attachments/assets/0969b814-fdbd-4501-8179-9254c90a4d06)

I also tried to configure the app via executable and got the same error:
```
builder.AddExecutable("frontend", "npm run dev", "../../dm-studio-frontend")
    .WithHttpEndpoint(port: 30010, env: "PORT")
    .WithReference(backend)
    .PublishAsDockerFile();

```
I am not sure if this is a runtime/aspire issue but if I just run the app from the terminal, it works. It only gives this error while running via aspire.

Some extra info:
IDE: Rider

dotnet version: 
`8.0.401`

aspire workload version: 
`aspire                     8.2.0/8.0.100         SDK 8.0.400`

ubuntu:
`24.04.1 LTS (Noble Numbat)`

Thanks in advance!

**Comment by davidfowl** [comment]
Did you install nodejs?

**Comment by adamtrip** [comment]
yes, the react app, in standalone, works just fine

**Comment by jack775544** [comment]
Did you install nodejs/npm via nvm? If you did then nvm does not add node/npm to the path but instead works by using a script in the bashrc to add the installation to the path on starting your terminal.

If this is what is happening to you, you would need to add the binaries to your system path (I believe in `~/.profile`) then it should work. Other option would be to start Rider from a terminal terminal and then it will inherit the PATH from that terminal.

**Comment by adamtrip** [comment]
@jack775544 as a matter of fact, I actually used fnm (alternative to nvm). is this a situation exclusive to linux? I usually use nvm on windows and never had this particular issue

**Comment by jack775544** [comment]
I've never used fnm before but a quick skim of the readme makes me feel like it would have the same issue.

NVM for Windows actually works a bit different to the Macos/Linux version. It creates a symlink in a folder (normally C:\Program Files\nodejs) and adds that folder to the system PATH. When you change versions of node, it will adjust the location that the symlink points to. This means that it doesn't rely on any shell script magic to add node to the path and causes everything to work normally.

If you want to test if this hypothesis is correct though, try and start the app host project with `dotnet run` from the terminal. If it can find npm, then you know it is a PATH related issue.

**Comment by dbreshears** [comment]
Appears this is a machine configuration issue, converting to discussion

**Comment by fdonnet** [comment]
confirmed not working with node installed with fnm (on Windows).

**Reply by davidfowl** [reply]
fnm ? Where does that put node?

**Reply by fdonnet** [reply]
i think something like that %userprofile%\AppData\Roaming\fnm\node-versions\vX.xx\installation

On windows, we add this line:
fnm env --use-on-cd --shell powershell | Out-String | Invoke-Expression

In the powershell profile (Invoke-Item $profile) to be able to switch via their "use" shortcut.

Cannot be extra sure about the final binary path, first time I used it because I needed 2 versions of node for different prj. 

But it's more something on our side to know that it has an impact on Aspire to proceed like that (with a node version switcher) and that it will break the "link". But when you try to run AddNpmApp it's hard to spot this issue at first and you will search for other "Aspire" issue when it's not really one ;).

**Comment by m0x6A** [comment]
I can also confirm this issue and probable cause. I tried running   [AspireWithNode](https://github.com/dotnet/aspire-samples/tree/main/samples/AspireWithNode) from the samples both in Rider and in the terminal. It finds node when running dotnet run through the terminal but not when starting the project through Rider.

**Comment by maftieu** [comment]
It seems that using `corepack yarn run ...` instead of `yarn run ...` it works with fnm (it may also work with nvm, but I cannot test, I use fnm).
Corepack is available from path, while yarn is'nt always.

```cs
public static IResourceBuilder<NodeAppResource> AddYarnApp(this IDistributedApplicationBuilder builder, [ResourceName] string name, string workingDirectory, string scriptName = "start", string[]? args = null)
{
	ArgumentNullException.ThrowIfNull(builder);
	ArgumentNullException.ThrowIfNull(name);
	ArgumentNullException.ThrowIfNull(workingDirectory);
	ArgumentNullException.ThrowIfNull(scriptName);
	string[] allArgs = args is { Length: > 0 }
		? ["yarn", "run", scriptName, "--", .. args]
		: ["yarn", "run", scriptName];

	workingDirectory = PathNormalizer.NormalizePathForCurrentPlatform(Path.Combine(builder.AppHostDirectory, workingDirectory));
	var resource = new NodeAppResource(name, "corepack", workingDirectory);

	return builder.AddResource(resource)
				  .WithNodeDefaults()
				  .WithArgs(allArgs);
}
```

Should be the same with YarnPackageInstaller
